### PR TITLE
Precompiles E2E: HSCS-PREC-008

### PR DESCRIPTION
**Pull request by georgiyazovaliiski**
**Description**:
This PR includes two changes:
- enables contract accounts to be able to collect custom hBar fees from token transfers
- implements HSCS-PREC-008: Account that would have otherwise insufficient balance able to send hbar received from custom fee payment figured by previous precompile.

### CODE REVIEW

File: test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/ContractHTSSuite.java [hunk 1]
```diff
@@ -478,6 +481,85 @@ private HapiApiSpec tokenTransferFromFeeCollector() {
...
+				).when(
+						getAccountBalance(CONTRACT).hasTinyBars(0),
+						withOpContext(
+								(spec, opLog) -> {
+									final var token = asAddress(spec.registry().getTokenID(A_TOKEN));
+									final var sender = asAddress(spec.registry().getAccountID(ACCOUNT));
+									final var receiver = asAddress(spec.registry().getAccountID(RECEIVER));
+									final var feeCollector = asAddress(spec.registry().getAccountID(FEE_COLLECTOR));
+
+									/* --- HSCS-PREC-009 --- */
```
Comment (stoqnkpL) [inline test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/ContractHTSSuite.java:532]: This should be /* --- HSCS-PREC-008 --- */

File: test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/ContractHTSSuite.java [hunk 2]
```diff
@@ -478,6 +481,85 @@ private HapiApiSpec tokenTransferFromFeeCollector() {
...
+													token,
+													sender,
+													receiver,
+													10L,
+													feeCollector
+											)
+													.payingWith(ACCOUNT)
+													.gas(48_000)
+													.via("distributeTx")
+													.alsoSigningWithFullPrefix(FEE_COLLECTOR)
```
Comment (stoqnkpL) [inline test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/ContractHTSSuite.java:545]: We should not need the FEE COLLECTOR signature, I think we can remove it

File: test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/ContractHTSSuite.java [hunk 3]
```diff
@@ -478,6 +481,85 @@ private HapiApiSpec tokenTransferFromFeeCollector() {
...
+													.hasKnownStatus(SUCCESS));
+								})
+
+				).then(
+						childRecordsCheck("distributeTx", SUCCESS,
+								recordWith()
+										.status(SUCCESS)
+										.tokenTransfers(
+												changingFungibleBalances()
```
Comment (stoqnkpL) [inline test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/ContractHTSSuite.java:554]: In the child rec we also have `assessed_custom_fees` and `transferList`, can we also assert their correct values?

File: test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/ContractHTSSuite.java [hunk 4]
```diff
@@ -478,6 +481,85 @@ private HapiApiSpec tokenTransferFromFeeCollector() {
...
+														.including(A_TOKEN, ACCOUNT, -10L)
+														.including(A_TOKEN, RECEIVER, 10L)
+										)),
+						getAccountBalance(CONTRACT).hasTinyBars(0),
+						getAccountBalance(FEE_COLLECTOR).hasTinyBars(1000000100)
```
Comment (stoqnkpL) [inline test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/ContractHTSSuite.java:559]: This should be 100.

File: test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/ContractHTSSuite.java [hunk 5]
```diff
@@ -478,6 +481,85 @@ private HapiApiSpec tokenTransferFromFeeCollector() {
 				);
 	}
 
+	private HapiApiSpec hbarTransferFromFeeCollector() {
+		return defaultHapiSpec("HbarTransferFromFeeCollector")
+				.given(
+						cryptoCreate(ACCOUNT).balance(10 * ONE_HUNDRED_HBARS).maxAutomaticTokenAssociations(10),
+						cryptoCreate(RECEIVER).maxAutomaticTokenAssociations(10),
+						cryptoCreate(TOKEN_TREASURY),
+						cryptoCreate(FEE_COLLECTOR),
```
Comment (stoqnkpL) [inline test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/ContractHTSSuite.java:490]: We should create the FEE_COLLECTOR account with intial balance of 0, this way we can assert it has a balance ot 100 at the end of the test; 1000000100 is not very clear

**Review by stoqnkpL — CHANGES_REQUESTED** [review]
Overall looks good but I think some more changes are required.
⚠️ There's a failing test in CryptoCreateSuite (xferRequiresCrypto()) that causes ci-cd to fail. It's not failing in `hts-contracts` branch.

**Review by stoqnkpL — APPROVED** [review]
LGTM

**Review by shemnon — APPROVED** [review]
LGTM

Curious as to why the gas prices went up in the EET tests as there were no gas schedule changes.  Is it just to have them all be the same?

**Comment by stoqnkpL** [comment]
> LGTM
> 
> Curious as to why the gas prices went up in the EET tests as there were no gas schedule changes. Is it just to have them all be the same?

Prices went up because the intrinsic gas PR got merged. And they weren't fixed there, because these EETs were part of other PRs that were merged right before/after that.

**Review by shemnon — APPROVED** [review]

**Review by tinker-michaelj — CHANGES_REQUESTED** [review]
Although this does let contracts be fee collectors, it **also** lets `ContractID`s be used directly in a `CryptoTransferTransactionBody`, which Product has previously disallowed. 

I'm double-checking which direction we want to go. 👍

**Review by tinker-michaelj — APPROVED** [review]
Confirmed with Product that we're now OK with using `ContractID`s in the `CryptoTransferTransactionBody`. 👍

**Pull request merged**